### ISSUE TITLE
Merging primary and clipboard into one single mechanism

### ISSUE THREAD

**Issue by fturco** [Keruspe/GPaste#136]
Is it possible to configure Gpaste and merge primary and clipboard into a single mechanism? I don't like seemingly unpredictable differences in behaviour when selecting text and pasting it with the middle mouse button and when explicitely using copy/paste commands. I'm not sure if this is handled by Gpaste or directly by the X server. I'm using Gpaste 3.18.2 on a Parabola GNU/Linux-libre system.

**Comment by Keruspe** [comment]
Hi,

Using the GPaste graphical user interface (either by typing <Ctrl><Alt>g or by typing `gpaste-client ui`, you can set a bunch a things in the "burger" menu on the top right corner. You'll probably find what you want there

**Comment by fturco** [comment]
I enabled the following options:

- Primary selection affects history
- Synchronize clipboard with primary selection

But it doesn't do what I want. I'm not sure which combination of option will do the trick. I unfortunately don't know how X or Gpaste work behind scenes, and toggling options at random is not the way to go.

**Comment by Keruspe** [comment]
By default, only things copied to clipboard are added to GPaste, with the first option you checked, things copied to the primary selection are also added to GPaste.

What the second option does is that, each time you copy something to either clipboard or primary selection, it's propagated to the other one.

Enabling those two options should definitely do what I thought you wanted.

Can you provide an example when it doesn't work as youd like it to?

**Comment by fturco** [comment]
The problem occurs when copying from Iceweasel to Emacs. Steps to reproduce:
1. Explicitely copy URL A from Iceweasel
2. Paste it in Emacs using `C-y`
3. Select another address (URL B) from Iceweasel
4. Paste it in Emacs using `C-y`
Result: there are two instances of URL A and no instance of URL B.
Expected result: one instance of URL A followed by an instance of URL B
I'm using Iceweasel 43.0.2 and Emacs 24.5.
I bet it's Emacs' fault but I'm still not sure.

**Comment by fturco** [comment]
I can obtain the desired behaviour by setting the following variables in Emacs:

* `x-select-enable-clipboard` to `t`
* `x-select-enable-primary` to `t`

For more info: http://www.emacswiki.org/emacs/CopyAndPaste
Not sure if this bug should be closed although.

**Comment by Keruspe** [comment]
Hmm, I will look if I find anything in that, but in the past, two pieces of software caused problems with clipboard handling (and especially primary selection handling actually): emacs and chromium.

**Comment by Keruspe** [comment]
Can you try with something else than emacs just to verify whether it works or not?

Also, I just pushed a small clipboard-watcher, if you can clone the master branch and build GPaste, you'll find it in bin/clipboard-watcher
It will print you each time clipboard or primary selection changes value

**Comment by fturco** [comment]
clipboard-watcher always prints nothing and quits with exit status 77.

**Comment by fturco** [comment]
I also tried some copy-and-paste between Iceweasel (Firefox) and Gedit. Steps to reproduce:

1. Explicitely copy an URL (URL A) from Iceweasel address bar
2. Explicitely paste URL A into Gedit (result: URL A will appear; OK)
3. Implicitely copy another URL (URL B) from Iceweasel address bar
4. Explicitely paste URL B into Gedit (result: URL A will appear; URL B expected)

So it seems the problem doesn't depend on Emacs.

**Comment by fturco** [comment]
I can reproduce the problem even without Iceweasel/Firefox. Just use Gedit. Write "hello world" and use "hello" instead of URL A and "world" instead of URL B.

**Comment by Keruspe** [comment]
Oh, yeah, you have to launch clipboard-watcher with `--dont-skip` sorry.

Will do some further testing, thanks for the test case. (I have to find a box running on X11 somewhere)

**Comment by Keruspe** [comment]
Ok, managed to get X working fine, enabled the options, now here is the output of clipboard-watcher, with some comments added to explain what's going on

```
# I select 'Hello'
PRIMARY changed: 'Hello'
# GPaste takes the ownership of 'Hello' which causes a new notification 
# The following notification as it's identical to the one before
PRIMARY changed: 'Hello'
# GPaste puts 'Hello' in clipboard because the option to sync clipboard and primary is on
CLIPBOARD changed: 'Hello'
# GPaste causes a new event to ensure it has the ownership, the event is discarded
CLIPBOARD changed: 'Hello'
# I explicitely select hitting ctrl-c
# GPaste discards the event as the contents is the same as the current one
CLIPBOARD changed: 'Hello'
# I unselect 'Hello' to go to the place I want to paste it, so there's nothing left in primary
PRIMARY changed: '(null)'
# I paste 'Hello' with ctrl-v   <<<============ Works fine
# I select 'world'
PRIMARY changed: 'world'
# GPaste takes the ownership of 'world' which causes a new notificaiton 
# The following notification as it's identical to the one before
PRIMARY changed: 'world'
# GPaste puts 'world' in clipboard because the option to sync clipboard and primary is on
CLIPBOARD changed: 'world'
# GPaste causes a new event to ensure it has the ownership, the event is discarded
CLIPBOARD changed: 'world'
# I paste 'world' with ctrl-v   <<<============ Works fine
# I unselect 'world'
PRIMARY changed: '(null)'
# I select 'Hello'
PRIMARY changed: 'Hello'
# GPaste takes the ownership of 'Hello' which causes a new notification 
# The following notification as it's identical to the one before
PRIMARY changed: 'Hello'
# GPaste puts 'Hello' in clipboard because the option to sync clipboard and primary is on
CLIPBOARD changed: 'Hello'
# GPaste causes a new event to ensure it has the ownership, the event is discarded
CLIPBOARD changed: 'Hello'
# I paste 'Hello' using middle click    <<<============ Works fine
# I unselect 'Hello'
PRIMARY changed: '(null)'
# I select 'world'
PRIMARY changed: 'world'
# GPaste takes the ownership of 'world' which causes a new notificaiton 
# The following notification as it's identical to the one before
PRIMARY changed: 'world'
# GPaste puts 'world' in clipboard because the option to sync clipboard and primary is on
CLIPBOARD changed: 'world'
# GPaste causes a new event to ensure it has the ownership, the event is discarded
CLIPBOARD changed: 'world'
# I paste 'world' using middle click    <<<============ Works fine
# I unselect 'world'
PRIMARY changed: '(null)'
```

Did you use exactly the same scenario? (I did all this in gedit as suggested)

**Comment by fturco** [comment]
I can reproduce a similar problem in an even simpler situation:

1. Open Gedit
2. Write "hello"
3. Select "hello"
4. Press Ctrl+c
5. Deselect "hello"
6. Click the middle mouse button

Result: nothing is pasted. Expected result: "hello" is pasted.

This is the corresponding output from `clipboard-watcher`:

```
./clipboard-watcher
PRIMARY changed: 'hello'
PRIMARY changed: 'hello'
CLIPBOARD changed: 'hello'
PRIMARY changed: '(null)'
```

**Comment by Keruspe** [comment]
Yep, this one I totally got, I pasted the whole thing to be sure that's what you were talking about.

In the case where you select something, then unselect it, you end up with nothing in the primary selection, indeed. That's the only case where it's desynchronised afaik.

Will fix this, 3.18.3 will probably be released tonight or tomorrow

**Comment by Keruspe** [comment]
This is now fixed and the fix will be part of 3.18.3

Thanks for reporting and for the details!

Please reopen if it still doesn't work with 3.18.3, but I'm confident at least this last one is fixed
